plugins/inputs/smart/smart.go

```diff
@@ -0,0 +1,299 @@
+	// 192 Power-Off_Retract_Count -O--C-   097   097   000    -    14716
+	attribute = regexp.MustCompile("^\\s*([0-9]+)\\s(\\S+)\\s+([-P][-O][-S][-R][-C][-K])\\s+([0-9]+)\\s+([0-9]+)\\s+([0-9]+)\\s+([-\\w]+)\\s+([\\w\\+\\.]+).*$")
+)
+
+type Smart struct {
+	Path     string
+	Nocheck  string
+	Excludes []string
+	Devices  []string
+}
+
+var sampleConfig = `
+  ## Optionally specify the path to the smartctl executable
+  # path = "/usr/bin/smartctl"
+  #
+  ## Skip checking disks in this power mode. Defaults to
+  ## "standby" to not wake up disks that have stoped rotating.
+  ## See --nockeck in the man pages for smartctl.
+  ## smartctl version 5.41 and 5.42 have faulty detection of
+  ## power mode and might require changing this value to
+  ## "never" depending on your disks.
+  # nocheck = "standby"
+  #
+  ## Optionally specify devices to exclude from reporting.
+  # excludes = [ "/dev/pass6" ]
+  #
+  ## Optionally specify devices and device type, if unset
+  ## a scan (smartctl --scan) for S.M.A.R.T. devices will
+  ## done and all found will be included except for the
+  ## excluded in excludes.
+  # devices = [ "/dev/ada0 -d atacam" ]
+`
+
+func (m *Smart) SampleConfig() string {
+	return sampleConfig
+}
+
+func (m *Smart) Description() string {
+	return "Read metrics from storage devices supporting S.M.A.R.T."
+}
+
+func (m *Smart) Gather(acc telegraf.Accumulator) error {
+	if len(m.Path) == 0 {
+		return fmt.Errorf("smartctl not found: verify that smartctl is installed and that smartctl is in your PATH")
+	}
+
+	devices := m.Devices
+	if len(devices) == 0 {
+		var err error
+		devices, err = m.scan()
+		if err != nil {
+			return err
+		}
+	}
+
+	errs := m.getAttributes(acc, devices)
+	if len(errs) > 0 {
+		var errStrs []string
+		for _, e := range errs {
+			errStrs = append(errStrs, e.Error())
+		}
+		return errors.New(strings.Join(errStrs, ", "))
+	}
+
+	return nil
+}
+
+// Scan for S.M.A.R.T. devices
+func (m *Smart) scan() ([]string, error) {
+
+	cmd := execCommand(m.Path, "--scan")
+	out, err := internal.CombinedOutputTimeout(cmd, time.Second*5)
+	if err != nil {
+		return []string{}, fmt.Errorf("failed to run command %s: %s - %s", strings.Join(cmd.Args, " "), err, string(out))
+	}
+
+	devices := []string{}
+	for _, line := range strings.Split(string(out), "\n") {
+		dev := strings.Split(line, "#")
+		if len(dev) > 1 && !excludedDev(m.Excludes, strings.TrimSpace(dev[0])) {
+			devices = append(devices, strings.TrimSpace(dev[0]))
+		}
+	}
+	return devices, nil
+}
+
+func excludedDev(excludes []string, deviceLine string) bool {
+	device := strings.Split(deviceLine, " ")
+	if len(device) != 0 {
+		for _, exclude := range excludes {
+			if device[0] == exclude {
+				return true
+			}
+		}
+	}
+	return false
+}
+
+// Get info and attributes for each S.M.A.R.T. device
+func (m *Smart) getAttributes(acc telegraf.Accumulator, devices []string) []error {
+
+	errchan := make(chan error)
```

Comment:
That's right, only use AddError or return.  This way Telegraf counts the correct number of errors in the internal plugin, and the logging looks right.